Sandbox contracts ws router

nicky-ru opened the pull request.

nicky-ru committed:
add operator registry and tests

nicky-ru committed:
Update projectregistrar.sol

nicky-ru committed:
add staking todo

nicky-ru committed:
add and test project registry

nicky-ru committed:
update operator registry

nicky-ru committed:
Update projectRegistry.test.ts

nicky-ru committed:
update project registry

nicky-ru committed:
update operator registry

nicky-ru committed:
update fleet manager

nicky-ru committed:
add custom errors

nicky-ru committed:
add operator custom errors

nicky-ru committed:
update project registry errors

nicky-ru committed:
add implementation of WSReceiver

nicky-ru committed:
add device registry implementation

nicky-ru committed:
Update WSReceiver.sol

nicky-ru committed:
update wsrouter

nicky-ru committed:
Update WSReceiver.sol

nicky-ru committed:
update tests

nicky-ru committed:
add helpers

nicky-ru committed:
Update projectregistrar.sol

nicky-ru committed:
Delete DeviceRegistry.sol

nicky-ru committed:
Delete IProjectRegistry.sol

nicky-ru committed:
update wsreceiver

nicky-ru committed:
Update FleetManager.sol

nicky-ru committed:
update ws router

nicky-ru committed:
Update OperatorRegistry.sol

nicky-ru committed:
Delete ProjectRegistry.sol

nicky-ru committed:
udpate tests

ququzone commented on smartcontracts/sandbox/contracts/interfaces/IRouter.sol:
```diff
@@ -33,21 +24,19 @@ interface IRouter {
+    function submit(
+        uint256 _projectId,
+        address _receiver,
+        bytes32 _batchMR,
```
_batchMR, devicesMR and _zkProof should packed into one `bytes _data` parameter, this can make Router more flexibility.

ququzone commented on smartcontracts/sandbox/contracts/W3bstreamRouter.sol:
```diff
@@ -21,40 +21,18 @@ contract W3bstreamRouter is IRouter, Initializable {
         fleetManager = _fleetManager;
     }
 
-    function register(uint256 _projectId, address _receiver) external override {
```
For security purpose, I think we maybe need project register their receiver first.

ququzone commented on smartcontracts/sandbox/contracts/interfaces/IFleetManager.sol:
```diff
@@ -4,7 +4,13 @@ pragma solidity ^0.8.19;
+    /// @param _node operator address
     /// @param _projectId project id
-    function isAllowed(address _operator, uint256 _projectId) external view returns (bool);
+    function isAllowed(address _node, uint256 _projectId) external view returns (bool);
```
According current sprout design, the address that submit data to Router is operator address, not node address. Node address will introduce to sprout, but not design currently.

ququzone commented:
I think we need to clearly the definitions for node and operator, there are below question to clarify:
1. node address is znode address or enode address?
2. operator address is means enode address?
@CoderZhi @nicky-ru @simonerom

simonerom commented:
> I think we need to clearly the definitions for node and operator, 
Regarding this, I'm unsure about how the final Sprout system will operate: specifically, whether the enode and znode will be managed by the same person/entity, or if they are part of two distinct networks. This distinction is important because it implies that there might be network rewards for znode operators as well as for enode operators.

> 1. node address is znode address or enode address?
> 2. operator address is means enode address?
>    @CoderZhi @nicky-ru @simonerom

My concept for the operator profiles is based on a single "W3bstream node operator." This role closely resembles the IoTeX delegate concept, owning a "profile" address. This address is the account responsible for creating and managing the operator profile, which includes a "node" address and a "reward" address. The "node" address is the account that actually generates the W3bstream blockchain transaction to the router contract, containing the zk proof and merkle roots. I believe this role best fits the enode. As for the znode, although we don't need to "trust" it (since we rely on the proof), if it is managed by a different entity, they might still wish to be acknowledged in terms of rewards. If this is the case, it seems that we should establish both enode and znode operator profiles. Additionally, it should be clearly defined which znode actually generated the proof submitted by the enode, for reward distribution purposes.

nicky-ru committed:
Update router

- submit() _data param combines batchMR, devicesMR and zkProof
- put back receiver registration

ququzone commented:
> > I think we need to clearly the definitions for node and operator,
> > Regarding this, I'm unsure about how the final Sprout system will operate: specifically, whether the enode and znode will be managed by the same person/entity, or if they are part of two distinct networks. This distinction is important because it implies that there might be network rewards for znode operators as well as for enode operators.
> 
> > 1. node address is znode address or enode address?
> > 2. operator address is means enode address?
> >    @CoderZhi @nicky-ru @simonerom
> 
> My concept for the operator profiles is based on a single "W3bstream node operator." This role closely resembles the IoTeX delegate concept, owning a "profile" address. This address is the account responsible for creating and managing the operator profile, which includes a "node" address and a "reward" address. The "node" address is the account that actually generates the W3bstream blockchain transaction to the router contract, containing the zk proof and merkle roots. I believe this role best fits the enode. As for the znode, although we don't need to "trust" it (since we rely on the proof), if it is managed by a different entity, they might still wish to be acknowledged in terms of rewards. If this is the case, it seems that we should establish both enode and znode operator profiles. Additionally, it should be clearly defined which znode actually generated the proof submitted by the enode, for reward distribution purposes.

I think we can only design enode smart contracts currently.
